src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -259,83 +259,121 @@ private JsonPropertyInfo GetPropertyWithUniqueAttribute(Type attributeType, Dict
-            if (PropertyCache.TryGetValue(stringPropertyName, out info))
+            if (!PropertyCache.TryGetValue(stringPropertyName, out JsonPropertyInfo info))
             {
-                // Check if we should add this to the cache.
-                // Only cache up to a threshold length and then just use the dictionary when an item is not found in the cache.
-                int count;
-                if (localPropertyRefsSorted != null)
-                {
-                    count = localPropertyRefsSorted.Length;
-                }
-                else
+                info = JsonPropertyInfo.s_missingProperty;
+            }
+
+            // Three code paths to get here:
+            // 1) info == s_missingProperty. Property not found.
+            // 2) key == info.PropertyNameKey. Exact match found.
+            // 3) key != info.PropertyNameKey. Match found due to case insensitivity.
+            Debug.Assert(info == null || key == info.PropertyNameKey || Options.PropertyNameCaseInsensitive);
+
+            // Check if we should add this to the cache.
+            // Only cache up to a threshold length and then just use the dictionary when an item is not found in the cache.
+            int cacheCount;
+            if (localPropertyRefsSorted != null)
+            {
+                cacheCount = localPropertyRefsSorted.Length;
+            }
+            else
+            {
+                cacheCount = 0;
+            }
+
+            // Do a quick check for the stable (after warm-up) case.
+            if (cacheCount < PropertyNameCountCacheThreshold)
+            {
+                // Do a slower check for the warm-up case.
+                if (frame.PropertyRefCache != null)
                 {
-                    count = 0;
+                    cacheCount += frame.PropertyRefCache.Count;
                 }
 
-                // Do a quick check for the stable (after warm-up) case.
-                if (count < PropertyNameCountCacheThreshold)
+                // Check again to append the cache up to the threshold.
+                if (cacheCount < PropertyNameCountCacheThreshold)
                 {
-                    // Do a slower check for the warm-up case.
-                    if (frame.PropertyRefCache != null)
+                    if (frame.PropertyRefCache == null)
                     {
-                        count += frame.PropertyRefCache.Count;
+                        frame.PropertyRefCache = new List<PropertyRef>();
                     }
 
-                    // Check again to append the cache up to the threshold.
-                    if (count < PropertyNameCountCacheThreshold)
-                    {
-                        if (frame.PropertyRefCache == null)
-                        {
-                            frame.PropertyRefCache = new List<PropertyRef>();
-                        }
-
-                        ulong key = info.PropertyNameKey;
```

Comment:
Note: this was using the key from the already cached value which may be different than the key calculated earlier from the JSON. This means case-insensitive caching wasn't working.